### PR TITLE
Fix out of bound error in system.remote_data_paths + disk web

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit b6773184150628f0307323694d8cc55083444301 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55468/b6773184150628f0307323694d8cc55083444301/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6470393625">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55468/b6773184150628f0307323694d8cc55083444301/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Disks/ObjectStorages/Web/WebObjectStorage.cpp [hunk 1]
```diff
@@ -162,7 +162,22 @@ bool WebObjectStorage::exists(const std::string & path) const
 
     if (startsWith(it->first, path)
         || (it != files.begin() && startsWith(std::prev(it)->first, path)))
+    {
+        shared_lock.unlock();
+        std::unique_lock unique_lock(metadata_mutex);
```
Comment (pufit) [inline src/Disks/ObjectStorages/Web/WebObjectStorage.cpp:167]: Why do we need `.unlock()` and `unique_lock` here?

Comment (alesapin) [reply src/Disks/ObjectStorages/Web/WebObjectStorage.cpp:167]: I'm not familiar with this code, but looks like yes. `exists` is such a function which on a good-path is read-only (need only shared_lock`), but on a bad path can modify state itself. So here we modify state similar to https://github.com/ClickHouse/ClickHouse/pull/55468/files#diff-64938a2d590929ddef45935da30a0458bcf7181c6585f34fdf7cebff96be3990R132-R144. That is why we releasing shared lock and acquiring write lock (unique_lock).

**Review by pufit — APPROVED** [review]
LGTM!

**Pull request merged**